src/index.ts

```diff
@@ -35,73 +36,109 @@ let traceAgent: StackdriverTracer;
-          undefined,
-      minorVersion: process.env.GAE_MINOR_VERSION || undefined
-    }
-  };
-
-  let envSetConfig: Config = {};
+function initConfig(userConfig: Forceable<Config>): Forceable<TopLevelConfig> {
+  let envSetConfig = {};
   if (!!process.env.GCLOUD_TRACE_CONFIG) {
     envSetConfig =
         require(path.resolve(process.env.GCLOUD_TRACE_CONFIG!)) as Config;
   }
-
-  // Internally, ignoreContextHeader is no longer being used, so convert the
-  // user's value into a value for contextHeaderBehavior. But let this value
-  // be overridden by the user's explicitly set value for contextHeaderBehavior.
-  const contextHeaderBehaviorUnderride = {
-    contextHeaderBehavior: projectConfig.ignoreContextHeader ? 'ignore' :
-                                                               'default'
-  };
-
   // Configuration order of precedence:
   // 1. Environment Variables
   // 2. Project Config
   // 3. Environment Variable Set Configuration File (from GCLOUD_TRACE_CONFIG)
   // 4. Default Config (as specified in './config')
-  const config = extend(
-      true, {[FORCE_NEW]: projectConfig[FORCE_NEW]}, defaultConfig,
-      envSetConfig, contextHeaderBehaviorUnderride, projectConfig, envConfig,
-      {plugins: {}});
-  // The empty plugins object guarantees that plugins is a plain object,
-  // even if it's explicitly specified in the config to be a non-object.
-
-  // Enforce the upper limit for the label value size.
-  if (config.maximumLabelValueSize >
-      Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT) {
-    config.maximumLabelValueSize = Constants.TRACE_SERVICE_LABEL_VALUE_LIMIT;
-  }
-  // Make rootSpanNameOverride a function if not already.
-  if (typeof config.rootSpanNameOverride === 'string') {
-    const spanName = config.rootSpanNameOverride;
-    config.rootSpanNameOverride = () => spanName;
-  } else if (typeof config.rootSpanNameOverride !== 'function') {
-    config.rootSpanNameOverride = (name: string) => name;
-  }
-
-  // If the CLS mechanism is set to auto-determined, decide now what it should
-  // be.
-  const ahAvailable = semver.satisfies(process.version, '>=8');
-  if (config.clsMechanism === 'auto') {
-    config.clsMechanism = ahAvailable ? 'async-hooks' : 'async-listener';
-  }
-
-  return config;
+  const mergedConfig: (typeof defaultConfig)&Forceable<Config> =
+      extend(true, {}, defaultConfig, envSetConfig, userConfig);
+  const forceNew = userConfig[FORCE_NEW];
+
+  const getInternalClsMechanism = (clsMechanism: string): TraceCLSMechanism => {
+    // If the CLS mechanism is set to auto-determined, decide now
+    // what it should be.
+    const ahAvailable = semver.satisfies(process.version, '>=8');
+    if (clsMechanism === 'auto') {
+      return ahAvailable ? TraceCLSMechanism.ASYNC_HOOKS :
+                           TraceCLSMechanism.ASYNC_LISTENER;
+    }
+    return clsMechanism as TraceCLSMechanism;
+  };
+  const getInternalMaximumLabelValueSize = (maximumLabelValueSize: number) =>
```

Comment:
nit: now that this is a single line function, it would be okay to move it inline. I'll leave it up to you though.